tests/test_utils.py

```diff
@@ -290,13 +299,381 @@ def test_random_interferometer(self):
+
+        with self.eng_sf:
+            sf.ops.Ket(initial_state) | (self.q_sf[0], self.q_sf[1])
+            S | self.q_sf[0]
+            B | self.q_sf
+            S | self.q_sf[1]
+            B | self.q_sf
+
+        with eng_extract:
+            S | q_extract[0]
+            B | q_extract
+            S | q_extract[1]
+            B | q_extract
+
+        # calculate final states for vectorize_mode= True
+        matrix_extract = extract_unitary(eng_extract, cutoff_dim=self.D, vectorize_modes=True)
+        final_state_extract = np.tile(matrix_extract.dot(initial_state.reshape([-1])), self.bsize)
+        final_state_sf = self.eng_sf.run().ket().reshape([-1])
+
+        # calculate final states for vectorize_mode=False
+        array_extract = extract_unitary(eng_extract, cutoff_dim=self.D, vectorize_modes=False)
+        final_array_extract = np.einsum("abcd,bd->ac", array_extract, initial_state)
+        final_array_sf = self.eng_sf.run().ket()
+
+        self.assertAllAlmostEqual(final_state_extract, final_state_sf, delta=self.tol)
+        self.assertAllAlmostEqual(final_array_extract, final_array_sf, delta=self.tol)
+
+    def test_extract_unitary_2_modes_tf(self):
+        """Test that unitary extraction works for 2 modes using the tf backend"""
+        self.logTestName()
+
+        eng_extract, q_extract = sf.Engine(num_subsystems=2)
+
+        S = sf.ops.Sgate(2)
+        B = sf.ops.BSgate(2.234, -1.165)
+
+        initial_state = np.complex64(np.random.rand(self.D, self.D) + 1j*np.random.rand(self.D, self.D))
+
+        with self.eng_sf:
+            sf.ops.Ket(initial_state) | (self.q_sf[0], self.q_sf[1])
+            S | self.q_sf[0]
+            B | self.q_sf
+            S | self.q_sf[1]
+            B | self.q_sf
+
+        with eng_extract:
+            S | q_extract[0]
+            B | q_extract
+            S | q_extract[1]
+            B | q_extract
+
+        # calculate final states for vectorize_mode= True
+        matrix_extract_tf_backend = extract_unitary(eng_extract, cutoff_dim=self.D, vectorize_modes=True, backend='tf')
+
+        with tf.Session() as sess:
+            sess.run(tf.global_variables_initializer())
+            final_state_tf_backend = sess.run(tf.einsum('ab,b', matrix_extract_tf_backend, tf.constant(initial_state.reshape([-1]))))
+
+        final_state_tf_backend = np.tile(final_state_tf_backend, self.bsize)
+        final_state_sf = self.eng_sf.run().ket().reshape([-1])
+
+        self.assertAllAlmostEqual(final_state_tf_backend, final_state_sf, delta=self.tol)
+
+    def test_extract_channel_2_modes(self):
+        """Test that channel extraction works for 2 modes"""
+        self.logTestName()
+
+        eng_extract, q_extract = sf.Engine(num_subsystems=2)
+
+        S = sf.ops.Sgate(2)
+        B = sf.ops.BSgate(2.234, -1.165)
+
+        initial_state = np.random.rand(self.D, self.D, self.D, self.D) + 1j*np.random.rand(self.D, self.D, self.D, self.D)
```

Comment:
fix initial state

Reply:
What's the right way of fixing the initial array if we want it to be quite large? Here it has D^4 complex numbers (1296 as D=6)

Reply:
You could use `np.random.seed(42)` directly above the call to `np.random.rand`. NumPy has a guarantee that it will produce the same result anywhere:

> Compatibility Guarantee A fixed seed and a fixed series of calls to ‘RandomState’ methods using the same parameters will always produce the same results up to roundoff error except when the values were incorrect.

See
1. https://stackoverflow.com/questions/40676205/cross-platform-numpy-random-seed
2. https://stackoverflow.com/questions/5836335/consistently-create-same-random-numpy-array

Reply:
Ah yes, that's a good idea. Is it sufficient to set it once for the whole test suite? Will it break/upset other parts of the tests?

Reply:
I set the seed after importing numpy

Reply:
:+1:

Thinking about it, probably the best place to put it is in the `setUp` method of each test class.